trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/remote/session/RemoteSessionImpl.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server.remote.session;
+
+import io.trino.s3.proxy.server.credentials.Credential;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.services.sts.StsClient;
+import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+class RemoteSessionImpl
+        implements RemoteSession
+{
+    private final StsAssumeRoleCredentialsProvider credentialsProvider;
+    private final StsClient stsClient;
+    private final Runnable closeFunction;
+    private final AtomicBoolean closed = new AtomicBoolean();
+
+    RemoteSessionImpl(StsAssumeRoleCredentialsProvider credentialsProvider, StsClient stsClient, Runnable closeFunction)
```

Comment:
all this class does with the `stsClient` is close. this and together with the `closeFunction` seem like a weird pattern. 
is there a simpler way of chaining closable objects? we can handle closing the `stsClient` in the `closeFunction` so this objects doesn't have to keep a reference to it.

Reply:
I removed `RemoteSessionImpl` and just made it an inner class in `RemoteSession`. wdyt?

Reply:
I don't see any changes to `RemoteSession`, did you push the change?

Reply:
@mosiac1 sorry - it's in `RemoteSessionController`.